gems/smithy-client/lib/smithy-client/codecs/cbor.rb

```diff
@@ -52,20 +56,35 @@ def format_data(value, shape)
         end
 
         def format_list(values, shape)
-          values.collect { |value| format_data(value, shape.member.shape) }
+          values.collect do |value|
+            next if value.nil? && !sparse?(shape)
+
+            if value.nil? && sparse?(shape)
+              nil
+            else
+              format_data(value, shape.member.shape)
+            end
+          end
         end
 
         def format_map(values, shape)
           values.each.with_object({}) do |(key, value), data|
-            data[key] = format_data(value, shape.value.shape)
+            next if value.nil? && !sparse?(shape)
+
+            data[key] =
+              if value.nil? && sparse?(shape)
+                nil
+              else
+                format_data(value, shape.value.shape)
+              end
           end
         end
 
         def format_structure(values, shape)
```

Comment:
I'm going to implement the (de)serialization of union shapes in a separate PR since I have some questions before implementing. Will be a small follow-up PR most likely.